package com.novelsbr.backend.exceptions;

import static org.junit.jupiter.api.Assertions.assertThrows;

import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.context.SpringBootTest;

import com.novelsbr.backend.domain.dto.AuthorDTO;
import com.novelsbr.backend.domain.entities.Author;
import com.novelsbr.backend.enums.UserRole;
import com.novelsbr.backend.repositories.AuthorRepository;
import com.novelsbr.backend.repositories.GenderRepository;
import com.novelsbr.backend.repositories.NovelRepository;
import com.novelsbr.backend.services.AuthorService;
import com.novelsbr.backend.services.NovelService;

@SpringBootTest
class AuthorExceptionsTest {
	
	@Autowired
	NovelRepository novelRepository;
	
	@Autowired
	AuthorRepository authorRepository;
	
	@Autowired
	GenderRepository genderRepository;
	
	@Autowired
	NovelService novelService;
	
	@BeforeEach
	void setUp() {
		novelRepository.deleteAll();
		genderRepository.deleteAll();
		authorRepository.deleteAll();
	}

	@Test
	void existingAuthorExceptionTest() {
		Author author = new Author(null, "JoÃ£o", "All Star", "joao@gmail.com", "1234", UserRole.AUTHOR);
		authorRepository.save(author);
		assertThrows(ExistingAuthorException.class, () -> novelService.findById(70L));
	}
	
	@Test
	void entityNotFoundException01() {
		assertThrows(entityNotFoundException.class, () -> authorService.save(new AuthorDTO(author)));
	}

}
